Mathlib/Data/Nat/Bitwise.lean

```diff
@@ -85,61 +74,67 @@ lemma bitwise_of_ne_zero {n m : Nat} (hn : n ≠ 0) (hm : m ≠ 0) :
   simp only [hn, hm, mod_two_iff_bod, ite_false, bit, bit1, bit0, Bool.cond_eq_ite]
   split_ifs <;> rfl
 
-@[simp]
-lemma bitwise'_of_ne_zero {n m : Nat} (hn : n ≠ 0) (hm : m ≠ 0) :
-    bitwise' f n m = bit (f (bodd n) (bodd m)) (bitwise' f (n / 2) (m / 2)) := by
-  conv_lhs => { rw [←bit_decomp n, ←bit_decomp m] }
-  rw [bitwise'_bit', bit, div2_val, div2_val]
-  case ham =>
-    obtain ⟨⟩ | n := n
-    · contradiction
-    · simp only [div2_succ, cond, bodd_succ, Bool.not_not]
-      cases bodd n <;> simp
-  case hbn =>
-    obtain ⟨⟩ | m := m
-    · contradiction
-    · simp only [div2_succ, cond, bodd_succ, Bool.not_not]
-      cases bodd m <;> simp
-
-lemma bitwise'_eq_bitwise (f) : bitwise' f = bitwise f := by
```

Comment:
I think perhaps we should keep this as
```lean
lemma bitwise_eq_binaryRec :
  bitwise f =
    binaryRec (fun n => cond (f false true) n 0) fun a m Ia =>
      binaryRec (cond (f true false) (bit a m) 0) fun b n _ => bit (f a b) (Ia n)
```
what do you think @ChrisHughes24?

Reply:
I agree

Reply:
I've brought back `bitwise_eq_binaryRec`. The proof is a bit messier than before, since I tried avoiding bringing back `bitwise'_of_ne` as an intermediary result, and instead used the lower level `binaryRec_of_ne_zero`.
```lean
lemma binaryRec_of_ne_zero {C : ℕ → Sort*} {z : C 0} {f n} (h : n ≠ 0) :
    binaryRec z f n = bit_decomp n ▸ f (bodd n) (div2 n) (binaryRec z f (div2 n))
```

I'd be curious to hear your thought about this proof. Is it OK, or should we bring back `bitwise'_of_ne_zero` (with the former definition of `bitwise'` inlined)?

Reply:
I think this is fine